upstairs/src/pseudo_file.rs

```diff
@@ -71,51 +71,47 @@ impl IOSpan {
         &mut self,
         block_io: &Arc<T>,
     ) -> Result<(), CrucibleError> {
-        block_io
+        let buffer = block_io
             .read(
                 Block::new(
                     self.affected_block_numbers[0],
                     self.block_size.trailing_zeros(),
                 ),
-                self.buffer.clone(),
+                Buffer::new(self.buffer.len()),
             )
-            .await
+            .await?;
+
+        self.buffer.eat(0, buffer);
+
+        Ok(())
     }
 
     #[instrument(skip(block_io))]
     pub async fn write_affected_blocks_to_volume<T: BlockIO>(
         &self,
         block_io: &Arc<T>,
     ) -> Result<(), CrucibleError> {
-        let bytes = Bytes::from(self.buffer.as_vec().await.clone());
-
         block_io
             .write(
                 Block::new(
                     self.affected_block_numbers[0],
                     self.block_size.trailing_zeros(),
                 ),
-                bytes,
+                self.buffer.as_bytes(),
             )
             .await
     }
 
     #[instrument]
     pub async fn read_from_blocks_into_buffer(&self, data: &mut [u8]) {
         assert_eq!(data.len(), self.sz as usize);
-
-        for (i, item) in data.iter_mut().enumerate() {
-            *item = self.buffer.as_vec().await[self.phase as usize + i];
```

Comment:
TIL that we were acquiring an async lock for each individual byte 😬

Reply:
:car: